File: 2025/day_1/c/main.c
#include <stdio.h>
#include <stdlib.h>

const int max = 100;
const int start = 50;

typedef struct rotation {
  char direction;
  int n;
} rotation;

// function headers
rotation *get_rotations(int *count);
int rotate_dial(rotation rotations[], int start, int count);

int main() {
  // get rotations array
  int count = 0;
  rotation *rotations = get_rotations(&count);
  // solve
  int ans = rotate_dial(rotations, start, count);

  // print answer
  printf("%d\n", ans);

  // free used memory
  free(rotations);

  return 0;
}

rotation *get_rotations(int *count) {
  FILE *file = fopen("../input.txt", "r");

  // rotations array
  rotation *rotations = malloc(sizeof(rotation) * 5000);

  char content[10];

  while (fgets(content, 10, file)) {
    struct rotation rotation;
    rotation.direction = content[0];
    rotation.n = atoi(&content[1]);

    rotations[(*count)++] = rotation;
  }

  fclose(file);
  return rotations;
}

int rotate_dial(rotation rotations[], int start, int count) {
  int pos = start;
  int zero_count = 0;
  for (int i = 0; i <= count; i++) {
    rotation rotation = rotations[i];
    switch (rotation.direction) {
    case 'L':
      pos -= (rotation.n % max);
      if (pos < 0) {
        pos += max;
      }
      break;
    case 'R':
      pos += (rotation.n % max);
      if (pos >= max) {
        pos -= max;
      }
      break;
    }
    if (pos == 0) {
      zero_count++;
    }
  }
  return zero_count;
}

File: 2025/day_1/go/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func get_rotations() []string {
	rotations, err := os.ReadFile("../input.txt")
	if err != nil {
		log.Fatalf("Error reading file: %v", err)
	}

	return strings.Split(string(rotations), "\n")

}

func rotate_dial_1(rotations []string, start int, max int) int {
	zero_count := 0
	pos := start

	for _, r := range rotations {
		if len(r) < 1 {
			break
		}
		d := r[:1]
		n, err := strconv.Atoi(r[1:])
		if err != nil {
			log.Fatalf("Error converting rotation to int: %v", err)
		}
		switch d {
		case "L":
			pos -= (n % max)
			if pos < 0 {
				pos += max
			}
		case "R":
			pos += (n % max)
			for pos >= max {
				pos -= max
			}

		}
		if pos == 0 {
			zero_count++
		}
	}

	return zero_count
}

func rotate_dial_2(rotations []string, start int, max int) int {
	zero_count := 0
	pos := start

	for _, r := range rotations {
		if len(r) < 1 {
			break
		}
		d := r[:1]
		n, err := strconv.Atoi(r[1:])
		if err != nil {
			log.Fatalf("Error converting rotation to int: %v", err)
		}
		switch d {
		case "L":
			if pos == 0 {
				zero_count--
			}
			pos -= n
			for pos < 0 {
				pos += max
				zero_count++
			}
		case "R":
			pos += n
			for pos >= max {
				pos -= max
				if pos != 0 {
					zero_count++
				}
			}

		}
		if pos == 0 {
			zero_count++
		}
	}

	return zero_count
}

func main() {
	rotations := get_rotations()
	start := 50
	max := 100
	ans := rotate_dial_1(rotations, start, max)
	fmt.Println("Part 1: ", ans)

	ans = rotate_dial_2(rotations, start, max)
	fmt.Println("Part 2: ", ans)
}
